### PR TITLE
CORDA-2704 - Allow Evolver to ignore carpentry errors

### PR DESCRIPTION
**Pull request by fenryka**
The logic being a carpentry failure is only fatal if the type we fail to
synthesise is required *post* evolution. If the evolver discards the
failed types then we'll never instantiate them and the synthesis
failure won't matter.

The Master / V4 changes for this issue are here https://github.com/corda/corda/pull/4856 made by @r3domfox Changes are being made seperatly as the V3 / V4 and above code paths have significantly diverged to the point a back-port is effectively a re-write.

### CODE REVIEW

**Comment by fenryka** [comment]
*NOTE* Not finished - needs more tests, needs a flag to disable feature

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt [hunk 1]
```diff
@@ -35,23 +34,54 @@ abstract class EvolutionSerializer(
...
-            if(resultsIndex >= 0) {
-                new[resultsIndex] = this
+    abstract class RemoteParam(var resultsIndex: Int, open val property: PropertySerializer) {
+        abstract fun readProperty(
+                obj: Any?,
+                schemas: SerializationSchemas,
+                input: DeserializationInput, new: Array<Any?>)
+    }
+
+    open class ValidRemoteParam(
```
[the remark this answers is not in the document]
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: hold over from some preior refactoring

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt [hunk 2]
```diff
@@ -35,23 +34,54 @@ abstract class EvolutionSerializer(
...
-     * @param property object to read the actual property value
+     * @property property object to read the actual property value
      */
-    data class OldParam(var resultsIndex: Int, val property: PropertySerializer) {
-        fun readProperty(obj: Any?, schemas: SerializationSchemas, input: DeserializationInput,
-                         new: Array<Any?>
-        ) = property.readProperty(obj, schemas, input).apply {
-            if(resultsIndex >= 0) {
-                new[resultsIndex] = this
+    abstract class RemoteParam(var resultsIndex: Int, open val property: PropertySerializer) {
```
[the remark this answers is not in the document]
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/EvolutionSerializer.kt]: good point, they're sucj a jotlinism I never think about them

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/DeserializationInput.kt [hunk 3]
```diff
@@ -107,7 +107,7 @@ class DeserializationInput(private val serializerFactory: SerializerFactory) {
                     is DescribedType -> {
                         // Look up serializer in factory by descriptor
                         val serializer = serializerFactory.get(obj.descriptor, schemas)
-                        if (SerializerFactory.AnyType != type && serializer.type != type && with(serializer.type) { !isSubClassOf(type) && !materiallyEquivalentTo(type) })
+                        if (SerializerFactory.AnyType != type && serializer.type != type && with(serializer.type) { !isSubClassOf(type) && !materiallyEquivalentTo(type) } && serializer.type != Uncarpntable::class.java)
```
Comment (r3domfox) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/DeserializationInput.kt]: typo: `Uncarpentable`

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializerFactory.kt [hunk 4]
```diff
@@ -21,6 +22,10 @@ import javax.annotation.concurrent.ThreadSafe
 data class SerializationSchemas(val schema: Schema, val transforms: TransformsSchema)
 data class FactorySchemaAndDescriptor(val schemas: SerializationSchemas, val typeDescriptor: Any)
 
+open class Environment {
```
Comment (r3domfox) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializerFactory.kt]: A bit dubious about this abstraction. Couldn't `environment` just be a lambda: `(String) -> String?`?

How about:

```
typealias Environment = (String) -> String?

class FeatureFlag(private val name: String) {
  companion object {
    val DISABLE_CORDA_2707 = FeatureFlag("DISABLE-CORDA-2707")
  }
  fun isSetIn(environment: Environment): Boolean = environment(name)?.toBoolean ?: false
}
```

then

```
class ThingUsingFeatureFlag(private val environment: Environment = System::getEnv) {
  fun thingControlledByFeatureFlag() {
    if (FeatureFlag.DISABLE_CORDA_2707.isSetIn(environment) { ... }
  }
}
```

as the condition?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/UncarpentableSerializer.kt [hunk 5]
```diff
@@ -14,7 +14,7 @@ class Uncarpntable
 class UncarpentableSerializer(private val name : String) : AMQPSerializer<Any> {
     private val error : String get() = "$name cannot be serialized"
 
-    override val type: Type get() = Uncarpntable::class.java
+    override val type: Type get() = Uncarepntable::class.java
```
Comment (r3domfox) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/UncarpentableSerializer.kt]: `Uncarpentable`, although I quite like "un-care-pntable", as it suggests a certain _laissez-faire_ attitude

**Pull request merged**